src/nvim/buffer.c

```diff
@@ -309,7 +309,7 @@ close_buffer (
   // buffers.
   // The caller must take care of NOT deleting/freeing when 'bufhidden' is
   // "hide" (otherwise we could never free or delete a buffer).
-  if (!buf->terminal) {
+  if (!buf->terminal && !buf->b_messages) {
```

Comment:
I think I segfaulted when trying to edit the buffer and a message came in .  But that was in the early stage.  Doesn't look like the problem still exists.

I removed all the `b_messages` checks and remembered the actual reason: The line highlights need to match the buffer lines.  It was easier to do because `bufhl_add_hl()` requires a highlight ID which I couldn't get without touching `:echohl` and a bunch of other functions.  I also thought it'd be kind of messed up if the user thought they controlled a buffer and lines started disappearing because the messages are being truncated.